types/react/index.d.ts

```diff
@@ -2282,6 +2281,11 @@ declare namespace React {
     }
 }
 
+type Defaultize<P, D = {}> =
+    & Pick<P, Exclude<keyof P, keyof D>>
+    & Partial<Pick<P, Extract<keyof P, keyof D>>>
+    & Partial<D>;
```

Comment:
So, a note on this. If both `defaultProps` and `propTypes` are present, this implies that something while only appears as a `defaultProp` and not a `propType` is OK. If `P` is actually specified (because the component has type definitions, for example), this means that defaultProps can silently augment the declared props in the public interface of the component.

Is that desirable? Ehhhhhhh. The alternative isn't elegant tho - pass in separate types (or have two flavors of Defaultize) for declared props vs inferred props into `Defaultize`, and if declaredProps is a sentinel value (doesn't exist), and neither do inferred props, then you want this intersection member (so it becomes a primary inference source), otherwise you probably don't.

Reply:
Yes, I think I'm going to change this whole thing too, in testing I found that naively intersecting the proper `Props` interface and the inferred `DefaultProps` will end up widening types for non-annotated prop types properties. Something like this is bad:
```tsx
interface Props {
  someEnum: 'a' | 'b' | 'c'
}
const propTypes = {
  someEnum: PropTypes.string.isRequired
}
class Foo extends Component<Props> {
  static propTypes = propTypes
}
const foo = <Foo /> // attributes['someEnum'] becomes type 'string'
```